### PR TITLE
docs, Update docs about testing locally provisioned clusters

### PR DESCRIPTION
**Pull request by oshoval**
This commit polish a bit the documentation on how to
test fresh provisioned VM based clusters locally.

* Add explanation on how to test gocli,
which is part of the provision tool-chain.
* Update example to use a more recent VM based provider version
(k8s-1.19)
* Add needed exports for cluster-up to ease the reader.
* Update gocli export to point to `quay.io` (CONTRIBUTING.md), as we switched to `quay.io`.

Note:
`KUBEVIRTCI_PROVISION_CHECK` was already mentioned in the docs.

Signed-off-by: Or Shoval <oshoval@redhat.com>

### CODE REVIEW

File: KUBEVIRT_LOCAL_TESTING.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ Steps:
```
Comment (dankenigsberg) [inline KUBEVIRT_LOCAL_TESTING.md:8]: the first sentence on the file is a bit out-of context. It is unclear to me what "the changes" is referring to. Can you rephrase it?
Do you mean something like "After making changes in a kubevirtci provider, you should test it locally before publishing it."  @dhiller ?

Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md:8]: > Do you mean something like "After making changes in a kubevirtci provider, you should test it locally before publishing it."

Yes i think this is the meaning, i will rephrase it, and lets see what Daniel thinks parallel to that.

Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md:12]: Updated, tried to keep both of the approaches of you and Daniel intact.

File: KUBEVIRT_LOCAL_TESTING.md [hunk 2]
```diff
@@ -1,34 +1,52 @@
...
+Lets go on the required steps, starting by provisioning, all the way to `make functest`.
 
 Steps:
 
 ### kubevirtci: provision cluster locally
 
 ```bash
 cd .../kubevirtci
-cd cluster-provision/k8s/1.17.0
+cd cluster-provision/k8s/1.19
```
Comment (dankenigsberg) [inline KUBEVIRT_LOCAL_TESTING.md:15]: in another PR: fix other instances of `git grep k8s/1....0/`

File: KUBEVIRT_LOCAL_TESTING.md [hunk 3]
```diff
@@ -1,34 +1,52 @@
...
 ```
 
 ### kubevirt: prepare for tests
 
 ```bash
-# set local provision test flag
+# set local provision test flag (mandatory)
 export KUBEVIRTCI_PROVISION_CHECK=1
-# sync changes to kubevirt cluster-up, notably cluster/images.sh
+# sync changes to kubevirt cluster-up, notably cluster/images.sh (needed in case kubevirt testing is needed as well)
```
Comment (dankenigsberg) [inline KUBEVIRT_LOCAL_TESTING.md:24]: I don't understand the comment, nor your added parenthesis.
I don't even understand what are the `...` on the `rsync` commandline.

Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md:24]: the `...` is the notation in this file to say "whatever the path before that"
so `rsync -av .../kubevirtci/cluster-up .../kubevirt/cluster-up`
means rsync between your kubevirtci/cluster-up folder to your kubevirt/cluster-up folder,
whatever those two paths parent dir is (this is the `...`)

i will update the whole line, because some of it is outdated (images.sh is deprecated)
and the _ci-configs/ is the thing that need to be synced in case kubevirt should be tested.

**Comment by oshoval** [comment]
Addressed the comment that are related to this PR.

File: KUBEVIRT_LOCAL_TESTING.md (hunk 2 above)
Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md:16]: https://github.com/kubevirt/kubevirtci/pull/578
didn't update the kind sriov, since the the provider still have 1.17.0 in its name all over.

File: KUBEVIRT_LOCAL_TESTING.md [hunk 4]
```diff
@@ -1,34 +1,53 @@
...
+```
+
+### start cluster (can be done in kubevirtci itself as well of course)                                                           
                                                                                           
 ```bash                                                                                   
 cd .../kubevirt                                                                           
-                                                                                          
+
+export KUBEVIRT_PROVIDER=k8s-1.19
+export KUBECONFIG=$(./cluster-up/kubeconfig.sh)
```
Comment (dankenigsberg) [inline KUBEVIRT_LOCAL_TESTING.md:41]: earlier you asked to `cd cluster-provision/k8s/1.19` which does not have `./cluster-up/kubeconfig.sh`

Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md:41]: right but you have `cd .../kubevirt` line 38
so now you do
and the comment that say `(can be done in kubevirtci itself as well of course)`
which in that case it means you need to cd .../kubevirtci

is it ok ?

File: KUBEVIRT_LOCAL_TESTING.md [hunk 5]
```diff
@@ -1,34 +1,53 @@
...
 
 Steps:
 
 ### kubevirtci: provision cluster locally
 
 ```bash
 cd .../kubevirtci
-cd cluster-provision/k8s/1.17.0
+cd cluster-provision/k8s/1.19
 ../provision.sh      # this also calls cluster-up to check whether the cluster will really start and have the pods ready
```
Comment (dankenigsberg) [inline KUBEVIRT_LOCAL_TESTING.md:17]: I don't follow. If this calls `cluster-up`, why do set the env variables only after?

Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md:17]: provision calls cluster-up just for sanity check the cluster it created, and then it cluster-down it.
The next paragraph is to spin your own cluster-up with the changes, test it deeper, 
or even test it with kubevirt (until this point we were for sure on kubevirtci, from this point we can move to kubevirt)

File: KUBEVIRT_LOCAL_TESTING.md (hunk 4 above)
Comment (dankenigsberg) [reply KUBEVIRT_LOCAL_TESTING.md:41]: but I don't really want to touch kubevirt, I only want to modify kubevirtci, and spin it up.

I understand that testing kubevirt is the main purpose of kubevirtci, but I think there should be a clear flow how to spin it up; and another addition how to sync kubevirt.

Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md:41]: i understand, but this is the purpose at least now of this document
and even its name
`KUBEVIRT_LOCAL_TESTING.md`

is it fine that this PR will be merged as is, as its just clarify the current flow that is presented in this PR.
and we create other md file which is kubevirtci only?

File: KUBEVIRT_LOCAL_TESTING.md (hunk 5 above)
Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md]: Added `KUBEVIRTCI_LOCAL_TESTING.md` which is kubevirtci dedicated.

**Comment by oshoval** [comment]
Added `KUBEVIRTCI_LOCAL_TESTING.md` which is kubevirtci dedicated.
Polished `KUBEVIRT_LOCAL_TESTING.md` due to this fact above.

File: KUBEVIRT_LOCAL_TESTING.md (hunk 4 above)
Comment (oshoval) [reply KUBEVIRT_LOCAL_TESTING.md:47]: Added KUBEVIRTCI_LOCAL_TESTING.md which is kubevirtci dedicated.

File: KUBEVIRTCI_LOCAL_TESTING.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
+# [kubevirtci](README.md): Testing kubevirtci with locally provisioned cluster
```
Comment (dankenigsberg) [inline KUBEVIRTCI_LOCAL_TESTING.md:1]: I think that adding this file is a bad idea: it has a big overlap with KUBEVIRT_LOCAL_TESTING (even the filename is almost indistinguishable). Any fix would have to be repeated twice and mismatch is bound to be created.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: KUBEVIRTCI_LOCAL_TESTING.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+After making changes to a kubevirtci provider, you should test it locally before publishing it.
+
+Let's go on the required steps, starting by provisioning, all the way to `make cluster-up`.
+Note: the `...` in the beginning of `.../directory_name` means the full path of `directory_name`.
```
Comment (dankenigsberg) [inline KUBEVIRTCI_LOCAL_TESTING.md:6]: Has `...` commonly used elsewhere? To me using `cd $KUBEVIRT_CI_DIR` would be more readable.

File: KUBEVIRTCI_LOCAL_TESTING.md (hunk 6 above)
Comment (oshoval) [reply KUBEVIRTCI_LOCAL_TESTING.md:1]: What about having the info on this new file KUBEVIRTCI_LOCAL_TESTING.md (and renaming it),
and that the previous file KUBEVIRT_LOCAL_TESTING.md will reference to the kubevirtci one, and just add the additional steps ?

Comment (dankenigsberg) [reply KUBEVIRTCI_LOCAL_TESTING.md:1]: I prefer to have one file, starting with pure kubevirtci testing, then a section with how to add kubevirt on top of it.

**Comment by oshoval** [comment]
Addressed comments
thanks

**Comment by oshoval** [comment]
Split to commits to ease review

File: KUBEVIRTCI_LOCAL_TESTING.md [hunk 8]
```diff
@@ -1,40 +1,90 @@
...
+```bash
+# switch directory to kubevirt folder
+cd $KUBEVIRT_DIR
+
+export KUBEVIRT_PROVIDER=k8s-1.19
+export KUBECONFIG=$(./cluster-up/kubeconfig.sh)
+export KUBEVIRT_NUM_NODES=2
+
+# spin up cluster
+make cluster-up
```
Comment (dankenigsberg) [inline KUBEVIRTCI_LOCAL_TESTING.md:76]: why is this needed? we already have a cluster ready from the former (kubevirtci-only) step

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: KUBEVIRTCI_LOCAL_TESTING.md (hunk 8 above)
Comment (oshoval) [reply KUBEVIRTCI_LOCAL_TESTING.md:76]: Theoretical its might not mandatory
but then you need to copy the `_ci-configs/` to `kubevirt` folder, because some portions of kubevirt e2e tests
are looking hardcoded in $KUBEVIRT_PROVIDER/_ci_configs

best also to export those two again
```
export KUBEVIRT_PROVIDER=k8s-1.19
export KUBECONFIG=$(./cluster-up/kubeconfig.sh)
```

I find it better to recreate a cluster, within the kubevirt folder, and test it end 2 end
instead doing the two hacks above

this way, it also make this step independent of the previous `cluster-up` step of kubevirtci
wdyt?

**Comment by oshoval** [comment]
Addressed comment (resolved the addressed one)

File: KUBEVIRTCI_LOCAL_TESTING.md [hunk 9]
```diff
@@ -1,40 +1,90 @@
...
 ### kubevirtci: provision cluster locally
 
 ```bash
-cd .../kubevirtci
-cd cluster-provision/k8s/1.17.0
-../provision.sh      # this also calls cluster-up to check whether the cluster will really start and have the pods ready
+cd $KUBEVIRTCI_DIR
+cd cluster-provision/k8s/1.19
+# this also calls cluster-up to check whether the cluster will really start and have the pods ready
+# but tear down the cluster afterwards, it's just a sanity check of the freshly provisioned cluster.
```
[the remark this answers is not in the document]
Comment (oshoval) [reply KUBEVIRTCI_LOCAL_TESTING.md]: Done

File: KUBEVIRTCI_LOCAL_TESTING.md (hunk 8 above)
Comment (dankenigsberg) [reply KUBEVIRTCI_LOCAL_TESTING.md:75]: > wdyt?

I don't no what is `ci_config` and I don't understand anything. How is the new cluster different from the old? As I see it, kubevirtci's `make cluster-up` + rsync + kubevirt's `make cluster-sync` should be enough. If you have to take the cluster down and use kubevirt's `make cluster-up` please explain what is the difference between the two `cluster-up`s.

Comment (oshoval) [reply KUBEVIRTCI_LOCAL_TESTING.md:75]: You don't need to cluster-down it, but you must copy `_ci_configs` folder to kubevirt,
else you will not be able to `cluster-sync` even,
this folder contains for example information about the ports of the cluster IIRC.
I just tried it.

I will add the `_ci_configs` to the rsync snippet and update the docs accordingly.

**Comment by oshoval** [comment]
Updated the info needed in order to reuse the cluster from kubevirtci part

**Comment by oshoval** [comment]
Added commit that remove the unneeded ginkgo prefix from ginkgo args

**Review by dankenigsberg — COMMENTED** [review]
I'm fine with the current version, but I'd like @dhiller (or someone else as knowledgeable) to approve.

**Comment by oshoval** [comment]
Thanks for the review.

**Comment by oshoval** [comment]
/cc @dhiller @fgimenez 
can you take a look please?
conceptual ACK from Dan https://github.com/kubevirt/kubevirtci/pull/576#pullrequestreview-623281419

File: KUBEVIRTCI_LOCAL_TESTING.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+Steps:
+
+### kubevirtci: provision cluster locally
+
+```bash
+cd $KUBEVIRTCI_DIR
+cd cluster-provision/k8s/1.19
+# Build a provider. This includes starting it with cluster-up for verification and shutting it down for cleanup.
+../provision.sh
```
Comment (dankenigsberg) [inline KUBEVIRTCI_LOCAL_TESTING.md:16]:
```suggestion
# Build a provider. This includes starting it with cluster-up for verification and shutting it down for cleanup.
(cd cluster-provision/k8s/1.19; ../provision.sh)
```
I prefer this, because nothing else needs to be run from `cluster-provision/k8s/1.19` dir

Comment (oshoval) [reply KUBEVIRTCI_LOCAL_TESTING.md]: Done

**Comment by oshoval** [comment]
Addressed comments
(2nd push was just to fix commit scope)

File: KUBEVIRTCI_LOCAL_TESTING.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+```bash
+# set local provision test flag (mandatory)
+export KUBEVIRTCI_PROVISION_CHECK=1
+```
+
+Note:
+In case you updated gocli and need to test it locally as well, export additionally:
+```bash
+export KUBEVIRTCI_GOCLI_CONTAINER=quay.io/kubevirtci/gocli:latest
```
Comment (dankenigsberg) [inline KUBEVIRTCI_LOCAL_TESTING.md:32]:
```suggestion
export KUBEVIRTCI_GOCLI_CONTAINER=quay.io/kubevirtci/gocli:latest
```
I cannot stand trailing whitespace

Comment (oshoval) [reply KUBEVIRTCI_LOCAL_TESTING.md]: Removed most, forgot 2
Done

Best to let sed do that xD

**Review by dankenigsberg — APPROVED** [review]
I managed to modify, build and start a provider based on this doc.

**Comment by oshoval** [comment]
/cc @dhiller @fgimenez
can you take a look please?

**Comment by oshoval** [comment]
/cc @qinqon 
mind taking a look please?
docs update on how to test fresh provisioned clusters locally

**Review by fgimenez — APPROVED** [review]
/lgtm
/approve

**Comment by oshoval** [comment]
Thanks

/retest

**Comment by oshoval** [comment]
can we override those please?
The PR is just a doc update
